File: 프로그래머스/lv1/92334. 신고 결과 받기/신고 결과 받기.java
import java.util.*;

class Solution {
    public int[] solution(String[] id_list, String[] report, int k) {
        int n = id_list.length;
        int[] answer = new int[n];
        Map<String, Integer> userIndex = new HashMap<>();
        Map<String, Set<String>> reporters = new HashMap<>();
        Map<String, Integer> reportCount = new HashMap<>();
        
        for (int i = 0; i < n; i++) {
            userIndex.put(id_list[i], i);
        }
        
        for (String rep : report) {
            String[] users = rep.split(" ");
            String reporter = users[0];
            String reported = users[1];
            
            if (!reporters.containsKey(reported)) {
                reporters.put(reported, new HashSet<>());
            }
            if (reporters.get(reported).add(reporter)) {
                reportCount.put(reported, reportCount.getOrDefault(reported, 0) + 1);
            }
        }
        
        for (String key : reportCount.keySet()) {
            if (reportCount.get(key) >= k) {
                Set<String> repUsers = reporters.get(key);
                for (String reporter : repUsers) {
                    int idx = userIndex.get(reporter);
                    answer[idx]++;
                }
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.py
def solution(id_list, report, k):
    score = {v: 0 for i, v in enumerate(id_list)}
    blowers = {v: set() for i, v in enumerate(id_list)}

    for i in report:
        blower, reported = i.split()[0], i.split()[1]
        if reported not in blowers[blower]:
            blowers[blower].add(reported)
            score[reported] += 1

    banned = [i for i in score.keys() if score[i] >= k]

    answer = []
    for i in range(len(id_list)):
        cnt = 0
        for j in banned:
            if j in blowers[id_list[i]]:
                cnt += 1
        answer.append(cnt)

    return answer